semantics/src/traversal/functions.rs

```diff
@@ -379,7 +377,11 @@ fn func_return(
         let attributes =
             expressions::expr_with_default_move(Rc::clone(&scope), Rc::clone(&context), value)?;
 
-        if attributes.typ != scope.borrow().func_return_type().into() {
+        let host_func_def = scope
+            .borrow()
+            .current_function_def()
+            .expect("Failed to get function definition");
+        if attributes.typ != host_func_def.return_type.into() {
```

Comment:
Same here. If we are checking a `return` and fail to lookup the function hosting the `return` then clearly we are in an unexpected situation, a bug that needs to be fixed.